### PR TITLE
feat(3012): add unit tests for remote trigger and stage trigger

### PR DESCRIPTION
**Pull request by y-oksaku**
## Context

This is part 2 of the additional functional testing of the trigger.

## Objective

This PR adds a unit test for remote triggers and stages.
This is because we want to make sure that existing trigger functionality is not broken so that we can safely proceed with the refactoring.

I add the following test.

```text
    ✔ Multiple [ a, b ] is triggered
    ✔ [ a, b, c ] is triggered
    ✔ [ ~a, a ] is triggered
    - [ ~a, a, b ] is triggered when a succeeds
    ✔ [ ~sd@1:a ] is triggered in a downstream
    ✔ [ ~sd@2:a ] is triggered in a upstream
    ✔ Multiple [ ~sd@1:a ] are triggered
    ✔ Multiple [ ~sd@2:a ] are triggered
    ✔ [ ~sd@1:a ] is triggered in multiple downstream
    ✔ [ ~sd@1:a ] is not triggered in a downstream when sd@1:a fails
    ✔ [ ~sd@2:a ] is not triggered in a upstream when sd@1:a fails
    ✔ [ ~sd@1:a ] is triggered in a downstream when sd@1:a fails once and then restarts and succeeds
    ✔ [ ~sd@2:a ] is triggered in a upstream when sd@2:a fails once and then restarts and succeeds
    - [ sd@1:a ] is triggered in a downstream
    ✔ [ sd@2:a ] is triggered in a upstream
    - Multiple [ sd@1:a ] are triggered in a downstream
    ✔ Multiple [ sd@2:a ] are triggered in a upstream
    - [ sd@1:a ] is triggered in multiple downstream
    - [ sd@1:a ] is not triggered in a downstream when sd@1:a fails
    ✔ [ sd@2:a ] is not triggered in a upstream when sd@2:a fails
    - [ sd@1:a ] is triggered in a downstream when sd@1:a fails once and then restarts and succeeds
    ✔ [ sd@2:a ] is triggered in a upstream when sd@2:a fails once and then restarts and succeeds
    ✔ [ ~sd@1:a, ~sd@1:b ] is triggered when sd@1:a succeeds
    ✔ [ ~sd@2:a, ~sd@2:b ] is triggered when sd@2:a succeeds
    ✔ [ ~sd@1:a, ~sd@1:b ] is triggered when sd@1:a fails and sd@1:b succeeds
    ✔ [ ~sd@2:a, ~sd@2:b ] is triggered when sd@2:a fails and sd@2:b succeeds
    ✔ [ ~sd@1:a, ~sd@1:b ] is triggered twice in a downstream when sd@1:a and sd@1:b succeed
    ✔ [ ~sd@2:a, ~sd@2:b ] is triggered once in a upstream when sd@2:a and sd@2:b succeed
    ✔ [ ~sd@1:a, ~b ] is triggered when sd@1:a succeeds
    ✔ [ ~sd@2:a, ~b ] is triggered when sd@2:a succeeds
    ✔ [ ~sd@1:a, ~b ] is triggered when b succeeds
    ✔ [ ~sd@2:a, ~b ] is triggered when b succeeds
    ✔ [ ~sd@1:a, ~b ] is triggered twice in downstream when sd@1:a and b succeed
    ✔ [ ~sd@2:a, ~b ] is triggered once in upstream when sd@2:a and b succeed
    ✔ [ sd@2:a, sd@2:b ] is not triggered when only sd@2:a succeeds
    ✔ [ sd@2:a, sd@2:b ] is not triggered when sd@2:a succeeds and sd@2:b fails
    ✔ [ sd@2:a, sd@2:b ] is triggered when sd@2:b fails once and then restarts and succeeds
    ✔ [ sd@2:a, sd@2:b ] is not triggered when sd@2:a fails and sd@2:b restarts and succeeds
    ✔ [ sd@2:a, sd@2:b ] is triggered
    ✔ Multiple [ sd@2:a, sd@2:b ] is triggered
    ✔ [ sd@2:c, sd@2:d ] is triggered (Multiple time Remote join)
    ✔ [ sd@2:a, b ] is triggered
    ✔ [ sd@2:a, b ] is not triggered when sd@2:a succeeds and b fails
    ✔ [ sd@2:a, b ] is triggered when sd@2:a succeeds and b fails and then restarts and succeeds
    ✔ [ sd@2:a, b ] is not triggered when b fails and sd@2:a restarts and succeeds
    ✔ [ sd@2:a, b ] is not triggered when sd@2:a fails and b succeeds
    ✔ [ sd@2:a, b ] is not triggered when sd@2:a fails and b restarts and succeeds
    ✔ [ sd@2:a, b ] is triggered when b succeeds and sd@2:a fails and then restarts and succeeds
    ✔ [ sd@2:a, sd@3:a ] is triggered
    ✔ [ sd@2:a, sd@3:a ] is not triggered when sd@2:a succeeds and sd@3:a fails
    ✔ [ sd@2:a, sd@3:a ] is not triggered when sd@3:a fails and sd@2:a restarts and succeeds 
    ✔ [ sd@2:a, sd@3:a ] is triggered when sd@2:a success and sd@3:a fails and then restarts and succeeds 
    ✔ [ ~sd@2:a, sd@2:b, sd@2:c ] is triggered when sd@2:a succeeds
    ✔ [ ~sd@2:a, sd@2:b, sd@2:c ] is triggered
    ✔ [ ~sd@2:a, sd@2:b, sd@2:c ] is triggered when sd@2:a fails and sd@2:b succeeds and sd@2:c succeeds
    ✔ [ ~sd@2:a, sd@2:b, sd@2:c ] is not triggered when sd@2:a fails and sd@2:b succeeds and sd@2:c fails
    ✔ [ ~sd@2:a, sd@2:b, sd@2:c ] is triggered when sd@2:a fails and sd@2:b succeeds and sd@2:c restarts and succeeds 
    ✔ [ ~sd@2:a, b, c ] is triggered
    ✔ [ ~sd@2:a, b, c ] is triggered when sd@2:a fails and b succeeds and c succeeds
    ✔ [ ~sd@2:a, b, c ] is not triggered when sd@2:a fails and b succeeds and c fails
    ✔ [ ~sd@2:a, sd@2:b, c ] is trigger
    ✔ [ ~sd@2:a, sd@2:b, c ] is trigger when sd@2:a fails and sd@2:b succeeds and c succeeds
    ✔ [ ~sd@2:a, sd@2:b, c ] is not trigger when sd@2:a fails and sd@2:b succeeds and c fails
    ✔ [ ~a, sd@2:b, sd@2:c ] is trigger
    ✔ [ ~a, sd@2:b, sd@2:c ] is trigger when a fails and sd@2:b succeeds and sd@2:c succeeds
    ✔ [ ~a, sd@2:b, sd@2:c ] is not trigger when a fails and sd@2:b succeeds and sd@2:c fails
    ✔ [ ~a, b, sd@2:c ] is trigger
    ✔ [ ~a, b, sd@2:c ] is trigger when a fails and b succeeds and sd@2:c succeeds
    ✔ [ ~a, b, sd@2:c ] is not trigger when a fails and b succeeds and sd@2:c fails
    ✔ [ sd@2:a, sd@2:b, sd@2:c ] is triggered
    ✔ [ ~sd@1:a, ~sd@1:a ] is triggered
    ✔ [ ~sd@1:a, sd@1:a ] is triggered
    - [ ~sd@1:a, sd@1:a, sd@1:b ] is triggered when a succeeds
    ✔ [ ~sd@2:a, ~sd@2:a ] is triggered
    ✔ [ ~sd@2:a, sd@2:a ] is triggered
    - [ ~sd@2:a, sd@2:a, sd@2:b ] is triggered when sd@2:a succeeds
    ✔ [ sd@3:a, sd@3:b ] is triggered in second stream pipeline
    ✔ [ sd@2:b, sd@3:a ] is triggered in first stream pipeline (Child and Grandchild Remote join)
    ✔ stage setup is triggered
    ✔ stage is triggered by stage
    ✔ stage jobs are triggered
    ✔ stage teardown is triggered
    ✔ [ ~stage@red:setup ] is triggered
    ✔ [ ~stage@red:teardown ] is triggered
    ✔ [ ~stage@red ] is triggered
    ✔ [ ~stage@red, ~stage@blue ] is triggered
```

Some tests are skipped since currently there are not implemented.

## References

Issue: https://github.com/screwdriver-cd/screwdriver/issues/3012

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: test/plugins/trigger.test.js [hunk 1]
```diff
@@ -876,147 +876,1466 @@ describe('trigger tests', () => {
...
-        // run all builds
-        await buildFactoryMock.run();
+        await upstreamEvent.run();
 
-        assert.equal(event.getBuildOf('target').status, 'SUCCESS');
+        const downstreamEvent = downstreamPipeline.getLatestEvent();
+
+        await downstreamEvent.getBuildOf('a').complete('FAILURE');
+
+        assert.isNull(upstreamEvent.getBuildOf('target'));
```
Comment (itleigns) [inline test/plugins/trigger.test.js:1006]: We also want to make sure that the build is not running on a new event.
```suggestion
        assert.equal(upstreamPipeline.getBuildsOf('target').length, 0);
```

File: test/plugins/trigger.test.js [hunk 2]
```diff
@@ -876,147 +876,1466 @@ describe('trigger tests', () => {
...
+        await upstreamEvent.getBuildOf('hub').complete('SUCCESS');
+        await upstreamEvent.getBuildOf('a').complete('FAILURE');
+        await upstreamEvent.getBuildOf('b').complete('SUCCESS');
+
+        const downstreamEvent = downstreamPipeline.getLatestEvent();
+
+        assert.equal(downstreamEvent.getBuildOf('target').status, 'RUNNING');
+    });
+
+    it('[ ~sd@2:a, ~sd@2:b ] is triggered when sd@2:a fails and sd@2:b fails', async () => {
```
Comment (itleigns) [inline test/plugins/trigger.test.js:1274]:
```suggestion
    it('[ ~sd@2:a, ~sd@2:b ] is triggered when sd@2:a fails and sd@2:b succeeds', async () => {
```

File: test/plugins/data/trigger/~sd@1:a_~b-downstream.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+shared:
+  image: node:20
+  steps:
+    - test: echo 'test'
+
+jobs:
+  b:
+    requires: [ ~sd@1:hub ]
+  target:
+    requires: [ ~sd@1:a, b ]
```
Comment (itleigns) [inline test/plugins/data/trigger/~sd@1:a_~b-downstream.yaml:10]:
```suggestion
    requires: [ ~sd@1:a, ~b ]
```

File: test/plugins/trigger.test.js [hunk 4]
```diff
@@ -876,147 +938,1691 @@ describe('trigger tests', () => {
...
+
+        await upstreamEvent.run();
+
+        const downstreamEvent = downstreamPipeline.getLatestEvent();
+
+        assert.equal(downstreamEvent.getBuildOf('target').status, 'RUNNING');
+        assert.equal(downstreamPipeline.getBuildsOf('target').length, 1);
+    });
+
+    xit('[ ~sd@1:a, sd@1:a, sd@1:b ] is triggered when a succeeds', async () => {
```
Comment (itleigns) [inline test/plugins/trigger.test.js:2349]:
```suggestion
    xit('[ ~sd@1:a, sd@1:a, sd@1:b ] is triggered when sd@1:a succeeds', async () => {
```

File: test/plugins/trigger.test.js [hunk 5]
```diff
@@ -876,147 +938,1691 @@ describe('trigger tests', () => {
...
+        const upstreamPipeline = await pipelineFactoryMock.createFromFile('~sd@1:a_sd@1:a_sd@1:b-upstream.yaml');
+        const downstreamPipeline = await pipelineFactoryMock.createFromFile('~sd@1:a_sd@1:a_sd@1:b-downstream.yaml');
+
+        const upstreamEvent = await eventFactoryMock.create({
+            pipelineId: upstreamPipeline.id,
+            startFrom: 'hub'
+        });
+
+        await upstreamEvent.run();
```
Comment (itleigns) [inline test/plugins/trigger.test.js:2358]: `sd@1:b` also starts.

**Comment by y-oksaku** [comment]
Thank you for your review 👍

File: test/plugins/trigger.test.js [hunk 6]
```diff
@@ -855,6 +873,50 @@ describe('trigger tests', () => {
...
+            startFrom: 'hub'
+        });
+
+        await event.getBuildOf('hub').complete('SUCCESS');
+        await event.getBuildOf('a').complete('SUCCESS');
+
+        assert.equal(event.getBuildOf('target').status, 'RUNNING');
+    });
+
+    xit('[ ~a, a, b ] is triggered when a succeeds', async () => {
```
Comment (tkyi) [inline test/plugins/trigger.test.js:906]:
```suggestion
    it('[ ~a, a, b ] is triggered when a succeeds', async () => {
```

**Review by tkyi — APPROVED** [review]

**Pull request merged**